Automatic releasing

andreogle opened the pull request:
Adds a workflow step that will automatically create a Git tag and Github release on pushes to `main` when the `version` in `package.json` has been incremented and there is no already existing tag. This is inline with some other API3 repos that follow this pattern, like the OEV auctioneer workflow.

Requires the following secrets to be set. @Siegrift it might be better if you use your credentials for these. The comments in the workflow explain what values to use

- [x] GH_USER_NAME
- [x] GH_USER_EMAIL
- [x] GH_ACCESS_TOKEN

Closes: https://github.com/api3dao/commons/issues/71

dcroote commented on scripts/tag-and-release.ts:
Thoughts on adapting this for export i.e. for https://github.com/api3dao/signed-api/issues/323 and https://github.com/api3dao/airseeker/issues/339? Having it in here in commons would be quite fitting to the repo's purpose and would avoid duplicating the script across repos.

Siegrift replied:
I think that's a very good idea.

Siegrift commented on .github/workflows/main.yml:
```diff
@@ -49,6 +60,47 @@ jobs:
     name: All required checks passed
     runs-on: ubuntu-latest
-    needs: [documentation, lint-build-test]
+    needs: [documentation, build-lint-test]
```
I think this will also need to be changed in repo settings after the merge.

andreogle replied:
What would need to be changed?

Siegrift reviewed (approved):
👍 LGTM

Siegrift commented on .github/workflows/main.yml:
```diff
@@ -1,12 +1,23 @@
+#    The following minimum permissions are required:
+#      Read - access to metadata
+#      Read & write - access to actions and code
+# 2. GH_USER_NAME - The name (not username) associated with the Git user. e.g. John Smith
```
I'm wondering if we should not use some dedicated "releasing" account for this. It feels a bit weird that the releases are done in the personal name.

Siegrift replied:
I think it's such a detail that is not worth changing. And the benefit is that it's going to point to the person owning the release (and likely the release token). We can move to a dedicated releasing account once there's need for it.

Siegrift reviewed (approved):
I've set the secrets to my name and email and added the access token.

andreogle commented:
@Siegrift @dcroote I've moved the `tag-and-release` script to the `src/` folder and added a README. I'm not entirely sure how to test this though until we actually release `commons` again with this change

Siegrift commented:
> I've moved the tag-and-release script to the src/ folder and added a README. I'm not entirely sure how to test this though until we actually release commons again with this change

I think we can just release a patch and see it in action.

Siegrift commented on README.md:
```diff
@@ -26,6 +26,7 @@ Read the documentation and sources of each module how to use it in the project.
 - [logger](./src/logger/README.md)
 - [processing](./src/processing/README.md)
 - [run-in-loop](./src/run-in-loop/README.md)
+- [scripts](./src/scripts/README.md)
```
`scripts` sounds too generic. What about `release-scripts`?

andreogle replied:
I figured something generic is actually what we wanted. I'm not attached to it though and can change to `release-scripts`

Siegrift replied:
I doubt we'll add more scripts so it seems better to make this more concrete.

Siegrift commented on src/scripts/tag-and-release.ts:
```diff
@@ -0,0 +1,103 @@
+  console.info(`Done!`);
+};
+
+main()
```
It's not good practice to export the module with side effect (executing the release). I think it should only export the feature and let the consumer call it.

andreogle replied:
This was also intentional as importing projects would all need to add their own `scripts` directory and copy-paste some boilerplate. I also don't feel strongly here though

Siegrift replied:
I think that pasting some boilerplate is inevitable. You need to edit the `.yml` file for the CI and define a target in scripts. At which point you already can have the `scripts` folder and be explicit in calling the release function from commons.

Siegrift commented on scripts/tag-and-release.ts:
Why not import the feature from the `src` folder? We could immediately test whether it works as expected.

Siegrift replied:
I mean that basically the same implementation is defined is `src/scripts/tag-and-release`. It would make sense to import that and use it in this script.

Siegrift reviewed (commented):
I've created https://github.com/api3dao/commons/issues/129 but it seems it will also be closed by this PR (and we should change add the "closes ..." command in the description).

andreogle commented on src/scripts/README.md:
```diff
@@ -0,0 +1,20 @@
+
+NOTE: It can only be run on the `main` branch.
+
+### Usage
```
Github only keeps old action runs around for [400 days](https://docs.github.com/en/rest/guides/using-the-rest-api-to-interact-with-checks?apiVersion=2022-11-28#retention-of-checks-data) and the repos that are currently using this are private (while `commons` is public).

I've updated the README to include all of the recommended steps though. Let me know what you think

Siegrift commented on src/release-scripts/README.md:
```diff
@@ -0,0 +1,83 @@
+# Scripts
```
# Release Scripts

Siegrift commented on src/release-scripts/README.md:
```diff
@@ -0,0 +1,83 @@
+import { tagAndRelease } from '@api3/commons'
+
+await tagAndRelease('my-repo-name'); // defaults to using the 'main' branch
+await tagAndRelease('my-repo-name', 'release-branch');
```
👍 Nice. I guess we can release this and plug it in right away to Auctioneer.

dcroote replied:
Yea thanks for doing this @andreogle! Will apply in https://github.com/api3dao/signed-api/issues/323 and https://github.com/api3dao/airseeker/issues/339

Siegrift reviewed (approved):
👍 LGTM. I'm looking forward to utilizing this.
